### ISSUE TITLE
self defined fair loss, then failed!

### ISSUE THREAD

**Issue by youwei2567** [microsoft/LightGBM#3794]
I try to use a customed loss, then I built one the same as fair loss. However the loss exploded, 

> [[LightGBM] [Warning] Using self-defined objective function
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 1.046216 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 376596
[LightGBM] [Info] Number of data points in the train set: 72912, number of used features: 1485
[LightGBM] [Warning] Using self-defined objective function
[0. 0. 0. ... 0. 0. 0.] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[1]	valid_0's l2: 20996.4
[174.29870857 195.10947016 235.29934377 ... 174.29870857 202.22508479
 167.00225998] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[2]	valid_0's l2: 933870
[ -497.54065909  -686.43496219 -1163.40611894 ...  -497.54065909
  -767.99736383  -437.85750774] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[3]	valid_0's l2: 3.01023e+09
[15002.57882654 26699.18925799 74777.61190547 ... 15002.57882654
 33826.68059502 11851.79677292] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[4]	valid_0's l2: 3.01023e+09
[15002.57882654 26699.18925799 74777.61190547 ... 15002.57882654
 33826.68059502 11851.79677292] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[5]	valid_0's l2: 3.01023e+09
[15002.57882654 26699.18925799 74777.61190547 ... 15002.57882654
 33826.68059502 11851.79677292] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[6]	valid_0's l2: 3.01023e+09
[15002.57882654 26699.18925799 74777.61190547 ... 15002.57882654
 33826.68059502 11851.79677292] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[7]	valid_0's l2: 3.01023e+09
[15002.57882654 26699.18925799 74777.61190547 ... 15002.57882654
 33826.68059502 11851.79677292] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[8]	valid_0's l2: 3.01023e+09
[15002.57882654 26699.18925799 74777.61190547 ... 15002.57882654
 33826.68059502 11851.79677292] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[9]	valid_0's l2: 3.01023e+09
[15002.57882654 26699.18925799 74777.61190547 ... 15002.57882654
 33826.68059502 11851.79677292] [61. 61. 73. ... 63. 69. 61.]
(72912,) (72912,)
(72912,) (72912,)
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[10]	valid_0's l2: 3.01023e+09

My code is down here:
`
import lightgbm as lgb
import numpy as np
params = {
    'boosting_type': 'gbdt',
    'metric': 'l2',
    'num_leaves': 31,
    'learning_rate': 0.05,
    'verbose': 1
}

def fr(preds, lgb_train):
    y_true = lgb_train.get_label()
    fair_c = 1.
    print(preds, y_true)
    x = preds - y_true
    print(x.shape, np.abs(x).shape)
    grad = fair_c * x / (np.abs(x) + fair_c)
    hess = fair_c ** 2 / (np.abs(x) + fair_c) ** 2
    print(grad.shape, hess.shape)
    return grad, hess 

gbm = lgb.train(params,
                lgb_train,
                num_boost_round=10,
                valid_sets=lgb_eval,  # eval training data
                fobj=fr
                )
`

As I try fair loss directly in source API, it worked well:

> [LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 1.030296 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 376596
[LightGBM] [Info] Number of data points in the train set: 72912, number of used features: 1485
[LightGBM] [Info] Start training from score 64.001673
[1]	valid_0's l2: 26.5549
[2]	valid_0's l2: 22.6731
[3]	valid_0's l2: 20.0473
[4]	valid_0's l2: 18.2441
[5]	valid_0's l2: 16.7925
[6]	valid_0's l2: 15.6746
[7]	valid_0's l2: 14.5084
[8]	valid_0's l2: 13.5685
[9]	valid_0's l2: 12.6618
[10]	valid_0's l2: 11.9169

this part code is down here:
`
import lightgbm as lgb
import numpy as np
params = {
    'boosting_type': 'gbdt',
    'metric': 'l2',
    'objective':'fair',
    'num_leaves': 31,
    'verbose': 1
}

gbm = lgb.train(params,
                lgb_train,
                num_boost_round=10,
                valid_sets=lgb_eval,  # eval training data
                )
`

**Comment by youwei2567** [comment]
is anything wrong in my self defined loss code? or is there anything i didnt notice in preprocess when using a self defined loss?

**Comment by youwei2567** [comment]
Checked!
Found that there is a slight difference in these two logs, "init_score".
Then I checked

-  params descriptor html 'https://github.com/microsoft/LightGBM/blob/master/docs/Parameters.rst'
-  same problems in  issue 3532 'https://github.com/microsoft/LightGBM/issues/3532#issuecomment-723714067'

Its the reason that why my custom fair loss got loss explosion while builtin fair loss didnt.

Then problem become to  'how can I use a non-artificial init_score in custom fair loss coz I didnt have one useful init_score file?'

**Comment by youwei2567** [comment]
Checked Again！
I found another same issue 3312 'https://github.com/microsoft/LightGBM/issues/3312#issuecomment-674955643' mentioned this question!
Then I'll  check whether I can reproduce the init score step.

**Comment by youwei2567** [comment]
Checked Againnnnn!

when build the dataset, we add step of init_score as follow:

`def init_score(y):
    return y - np.mean(y)

from sklearn.model_selection import train_test_split
import lightgbm as lgb
X_train, X_eval, y_train, y_eval = train_test_split(first_half, first_half_label[:, 2], test_size=0.2, random_state=0)
lgb_train = lgb.Dataset(X_train, init_score(y_train), free_raw_data=False)
lgb_eval = lgb.Dataset(X_eval, init_score(y_eval), reference=lgb_train, free_raw_data=False)`

it works!
